File: crm_final/lib/models/caller_details.dart
class CallerDetails {
  final int? id;
  final int callerId;
  final int tasksLate;
  final int tasksPending;
  final int tasksDone;
  final int tasksCreated;
  final int whatsappIncoming;
  final int whatsappOutgoing;
  final int stageFresh;
  final int stageInterested;
  final int stageCommitted;
  final int stageNotInterested;
  final int stageNotConnected;
  final int stageCallback;
  final int stageTempleVisit;
  final int stageTempleDonor;
  final int stageLost;
  final int stageWon;
  final String? createdAt;

  CallerDetails({
    this.id,
    required this.callerId,
    required this.tasksLate,
    required this.tasksPending,
    required this.tasksDone,
    required this.tasksCreated,
    required this.whatsappIncoming,
    required this.whatsappOutgoing,
    required this.stageFresh,
    required this.stageInterested,
    required this.stageCommitted,
    required this.stageNotInterested,
    required this.stageNotConnected,
    required this.stageCallback,
    required this.stageTempleVisit,
    required this.stageTempleDonor,
    required this.stageLost,
    required this.stageWon,
    this.createdAt,
  });

  factory CallerDetails.fromJson(Map<String, dynamic> json) {
    return CallerDetails(
      id: json['id'],
      callerId: json['caller_id'],
      tasksLate: json['tasks_late'],
      tasksPending: json['tasks_pending'],
      tasksDone: json['tasks_done'],
      tasksCreated: json['tasks_created'],
      whatsappIncoming: json['whatsapp_incoming'],
      whatsappOutgoing: json['whatsapp_outgoing'],
      stageFresh: json['stage_fresh'],
      stageInterested: json['stage_interested'],
      stageCommitted: json['stage_committed'],
      stageNotInterested: json['stage_not_interested'],
      stageNotConnected: json['stage_not_connected'],
      stageCallback: json['stage_callback'],
      stageTempleVisit: json['stage_temple_visit'],
      stageTempleDonor: json['stage_temple_donor'],
      stageLost: json['stage_lost'],
      stageWon: json['stage_won'],
      createdAt: json['created_at'],
    );
  }

  Map<String, dynamic> toJson() {
    return {
      'id': id,
      'caller_id': callerId,
      'tasks_late': tasksLate,
      'tasks_pending': tasksPending,
      'tasks_done': tasksDone,
      'tasks_created': tasksCreated,
      'whatsapp_incoming': whatsappIncoming,
      'whatsapp_outgoing': whatsappOutgoing,
      'stage_fresh': stageFresh,
      'stage_interested': stageInterested,
      'stage_committed': stageCommitted,
      'stage_not_interested': stageNotInterested,
      'stage_not_connected': stageNotConnected,
      'stage_callback': stageCallback,
      'stage_temple_visit': stageTempleVisit,
      'stage_temple_donor': stageTempleDonor,
      'stage_lost': stageLost,
      'stage_won': stageWon,
      'created_at': createdAt,
    };
  }
}

File: crm_backend/models/caller_details.js
const db = require('../db');

exports.create = async (data) => {
    const [result] = await db.query(
        `INSERT INTO caller_details (
            caller_id, tasks_late, tasks_pending, tasks_done, tasks_created,
            whatsapp_incoming, whatsapp_outgoing,
            stage_fresh, stage_interested, stage_committed, stage_not_interested, stage_not_connected, stage_callback, stage_temple_visit, stage_temple_donor, stage_lost, stage_won
        ) VALUES (?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?)`,
        [
            data.caller_id,
            data.tasks_late || 0,
            data.tasks_pending || 0,
            data.tasks_done || 0,
            data.tasks_created || 0,
            data.whatsapp_incoming || 0,
            data.whatsapp_outgoing || 0,
            data.stage_fresh || 0,
            data.stage_interested || 0,
            data.stage_committed || 0,
            data.stage_not_interested || 0,
            data.stage_not_connected || 0,
            data.stage_callback || 0,
            data.stage_temple_visit || 0,
            data.stage_temple_donor || 0,
            data.stage_lost || 0,
            data.stage_won || 0
        ]
    );
    return result.insertId;
};

exports.findAll = async () => {
    const [rows] = await db.query('SELECT * FROM caller_details');
    return rows;
};

exports.findById = async (id) => {
    const [rows] = await db.query('SELECT * FROM caller_details WHERE id = ?', [id]);
    return rows[0] || null;
};

exports.findByCallerId = async (caller_id) => {
    const [rows] = await db.query('SELECT * FROM caller_details WHERE caller_id = ?', [caller_id]);
    return rows;
};

exports.update = async (id, data) => {
    const fields = [];
    const values = [];
    const columns = [
        'tasks_late', 'tasks_pending', 'tasks_done', 'tasks_created',
        'whatsapp_incoming', 'whatsapp_outgoing',
        'stage_fresh', 'stage_interested', 'stage_committed', 'stage_not_interested', 'stage_not_connected', 'stage_callback', 'stage_temple_visit', 'stage_temple_donor', 'stage_lost', 'stage_won'
    ];
    columns.forEach(col => {
        if (data[col] !== undefined) {
            fields.push(`${col} = ?`);
            values.push(data[col]);
        }
    });
    if (fields.length === 0) return;
    values.push(id);
    await db.query(`UPDATE caller_details SET ${fields.join(', ')} WHERE id = ?`, values);
};

exports.delete = async (id) => {
    await db.query('DELETE FROM caller_details WHERE id = ?', [id]);
}; 